lib/spack/llnl/util/filesystem.py

```diff
@@ -163,22 +164,28 @@ def getuid():
         return os.getuid()
 
 
-@system_path_filter
+@pathlib_filter
 def rename(src, dst):
     # On Windows, os.rename will fail if the destination file already exists
     if sys.platform == "win32":
         # Windows path existence checks will sometimes fail on junctions/links/symlinks
-        # so check for that case
-        if os.path.exists(dst) or os.path.islink(dst):
-            os.remove(dst)
-    os.rename(src, dst)
+        # check for that case
+        if dst.exists() or dst.is_symlink():
+            # unlink will fail for directoires
+            # use rmdir in that case
+            dst.unlink() if dst.is_file() else dst.rmdir()
+    src.rename(dst)
 
 
-@system_path_filter
+@pathlib_filter
 def path_contains_subdirectory(path, root):
-    norm_root = os.path.abspath(root).rstrip(os.path.sep) + os.path.sep
-    norm_path = os.path.abspath(path).rstrip(os.path.sep) + os.path.sep
-    return norm_path.startswith(norm_root)
+    norm_root = root.resolve()
+    norm_path = path.resolve()
+    try:
+        PurePath(norm_path).relative_to(norm_root)
+        return True
+    except ValueError:
+        return False
```

Comment:
Could we use [`.is_relative_to()`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to) here?

```suggestion
    return PurePath(norm_path).is_relative_to(norm_root)
```

Reply:
Unfortunately that only became available in 3.9 and IIRC we still want to support 3.7 & 3.8.

Reply:
We could always backport, it would be _really_ nice to have that method

Reply:
I'll consider making a separate PR with a compat method for 3.7 and 3.8; we can leave this as-is!